File: sdk-python/obsvr/strict_receipt_v2.py
"""Versioned strict receipts with signed tenant and target-hash binding."""

from __future__ import annotations

import base64
import binascii
import hashlib
from typing import Any, Dict, List, Optional

from .intent_alignment_v2 import INTENT_V2_ENGINE_VERSION
from .strict_canonical import (
    STRICT_IDENTIFIER_MAX_BYTES,
    STRICT_SET_MAX_ITEMS,
    bounded_canonical_text,
)
from .strict_receipt import (
    STRICT_RECEIPT_PROFILE_VERSION,
    STRICT_RECEIPT_SCHEMA,
    build_strict_receipt_body,
)
from .tool_pinning import _canonical_json_for_hash

STRICT_RECEIPT_V2_SCHEMA = "obsvr-strict-receipt-v2"
STRICT_RECEIPT_V2_PROFILE_VERSION = "2.0"
STRICT_RECEIPT_V2_ENVELOPE_SCHEMA = "obsvr-strict-receipt-envelope-v2"
STRICT_RECEIPT_V2_BODY_DOMAIN = b"obsvr-strict-receipt/body/2"
STRICT_RECEIPT_V2_SIGNATURE_DOMAIN = b"obsvr-strict-receipt/signature/2"

_HEX = frozenset("0123456789abcdef")


class StrictReceiptV2ValidationError(ValueError):
    """The value cannot be represented as one strict v2 receipt."""


def _fail(message: str) -> None:
    raise StrictReceiptV2ValidationError(message)


def _record(value: Any, field: str) -> Dict[str, Any]:
    if not isinstance(value, dict):
        _fail(f"{field} must be an object")
    return value


def _exact(value: Dict[str, Any], allowed: set[str], field: str) -> None:
    unknown = sorted(set(value) - allowed)
    if unknown:
        _fail(f"{field} contains unsupported field: {unknown[0]}")


def _text(value: Any, field: str) -> str:
    return bounded_canonical_text(value, field, STRICT_IDENTIFIER_MAX_BYTES, _fail)


def _hash(value: Any, field: str) -> str:
    if (
        not isinstance(value, str)
        or len(value) != 64
        or any(character not in _HEX for character in value)
    ):
        _fail(f"{field} must be 64 lowercase hex characters")
    return value


def _optional_hash(source: Dict[str, Any], key: str, field: str) -> Optional[str]:
    return _hash(source[key], field) if key in source else None


def _bounded_texts(values: Optional[List[str]], field: str) -> None:
    if values is None:
        return
    if len(values) > STRICT_SET_MAX_ITEMS:
        _fail(f"{field} exceeds {STRICT_SET_MAX_ITEMS} items")
    for index, value in enumerate(values):
        _text(value, f"{field}[{index}]")


def _bounded_input_texts(value: Any, field: str) -> None:
    if not isinstance(value, list):
        _fail(f"{field} must be an array")
    if len(value) > STRICT_SET_MAX_ITEMS:
        _fail(f"{field} exceeds {STRICT_SET_MAX_ITEMS} items")
    for index, item in enumerate(value):
        _text(item, f"{field}[{index}]")


def _u64(value: int) -> bytes:
    return value.to_bytes(8, "big", signed=False)


def strict_receipt_v2_key_id(raw_public_key: bytes) -> str:
    if not isinstance(raw_public_key, bytes) or len(raw_public_key) != 32:
        _fail("public key must be 32 raw bytes")
    return "sha256:" + hashlib.sha256(raw_public_key).hexdigest()


def _validate_text_bounds(body: Dict[str, Any]) -> None:
    for value, field in (
        (body["receipt_id"], "receipt_id"),
        (body["session_id"], "session_id"),
        (body["sdk"]["version"], "sdk.version"),
        (body["initiator"]["agent_id"], "initiator.agent_id"),
        (body["action"]["action_id"], "action.action_id"),
        (body["action"]["kind"], "action.kind"),
        (body["action"]["name"], "action.name"),
        (body["context"]["run_id"], "context.run_id"),
        (body["evaluation"]["engine_version"], "evaluation.engine_version"),
        (body["evaluation"]["policy_version"], "evaluation.policy_version"),
        (body["evaluation"]["reason_code"], "evaluation.reason_code"),
    ):
        _text(value, field)
    if "thread_id" in body["context"]:
        _text(body["context"]["thread_id"], "context.thread_id")
    _bounded_texts(body["evaluation"]["rule_ids"], "evaluation.rule_ids")
    suspension = body.get("suspension")
    if suspension:
        _text(suspension["suspension_id"], "suspension.suspension_id")
        _bounded_texts(suspension["required_fields"], "suspension.required_fields")
        if "approval_request_id" in suspension:
            _text(suspension["approval_request_id"], "suspension.approval_request_id")
    resolution = body.get("resolution")
    if resolution:
        _text(resolution["suspension_id"], "resolution.suspension_id")
        _text(resolution["resolver_principal_id"], "resolution.resolver_principal_id")


def build_strict_receipt_v2_body(input_value: Any) -> Dict[str, Any]:
    root = _record(input_value, "receipt body")
    _exact(
        root,
        {
            "schema",
            "profile_version",
            "record_type",
            "receipt_id",
            "tenant_id",
            "session_id",
            "sequence",
            "timestamp_ms",
            "clock_regression_clamped",
            "previous_receipt_hash",
            "sdk",
            "initiator",
            "action",
            "context",
            "evaluation",
            "execution_authorized",
            "suspension",
            "resolution",
        },
        "receipt body",
    )
    if root.get("schema") != STRICT_RECEIPT_V2_SCHEMA:
        _fail("invalid receipt schema")
    if root.get("profile_version") != STRICT_RECEIPT_V2_PROFILE_VERSION:
        _fail("invalid receipt profile_version")
    tenant_id = _text(root.get("tenant_id"), "tenant_id")
    raw_action = _record(root.get("action"), "action")
    _exact(
        raw_action,
        {
            "action_id",
            "kind",
            "name",
            "arguments_hash",
            "target_hash",
            "effective_arguments_hash",
        },
        "action",
    )
    target_hash = _optional_hash(raw_action, "target_hash", "action.target_hash")
    raw_context = _record(root.get("context"), "context")
    _exact(raw_context, {"schema", "context_hash", "run_id", "thread_id"}, "context")
    if raw_context.get("schema") != "obsvr-action-context-v2":
        _fail("invalid context.schema")
    raw_evaluation = _record(root.get("evaluation"), "evaluation")
    _bounded_input_texts(raw_evaluation.get("rule_ids"), "evaluation.rule_ids")
    if "suspension" in root:
        raw_suspension = _record(root["suspension"], "suspension")
        _bounded_input_texts(
            raw_suspension.get("required_fields"), "suspension.required_fields"
        )

    legacy_action = dict(raw_action)
    legacy_action.pop("target_hash", None)
    if target_hash is not None:
        legacy_action["target"] = target_hash
    legacy_input = {
        **root,
        "schema": STRICT_RECEIPT_SCHEMA,
        "profile_version": STRICT_RECEIPT_PROFILE_VERSION,
        "action": legacy_action,
        "context": {**raw_context, "schema": "obsvr-action-context-v1"},
    }
    legacy_input.pop("tenant_id")
    try:
        legacy = build_strict_receipt_body(legacy_input)
    except ValueError as error:
        raise StrictReceiptV2ValidationError(str(error)) from None
    normalized_action = dict(legacy["action"])
    normalized_target = normalized_action.pop("target", None)
    if normalized_target is not None:
        normalized_action["target_hash"] = normalized_target
    body = {
        **legacy,
        "schema": STRICT_RECEIPT_V2_SCHEMA,
        "profile_version": STRICT_RECEIPT_V2_PROFILE_VERSION,
        "tenant_id": tenant_id,
        "action": normalized_action,
        "context": {**legacy["context"], "schema": "obsvr-action-context-v2"},
    }
    _validate_text_bounds(body)
    if body["evaluation"]["engine_version"] != INTENT_V2_ENGINE_VERSION:
        _fail(f"evaluation.engine_version must be {INTENT_V2_ENGINE_VERSION}")
    return body


def canonicalize_strict_receipt_v2_body(input_value: Any) -> str:
    return _canonical_json_for_hash(build_strict_receipt_v2_body(input_value))


def strict_receipt_v2_hash(input_value: Any) -> str:
    body = canonicalize_strict_receipt_v2_body(input_value).encode("utf-8")
    preimage = STRICT_RECEIPT_V2_BODY_DOMAIN + b"\x00" + _u64(len(body)) + body
    return hashlib.sha256(preimage).hexdigest()


def strict_receipt_v2_signature_preimage(key_id: str, receipt_hash: str) -> bytes:
    if (
        not isinstance(key_id, str)
        or not key_id.startswith("sha256:")
        or len(key_id) != 71
        or any(character not in _HEX for character in key_id[7:])
    ):
        _fail("invalid strict key id")
    key = key_id.encode("utf-8")
    digest = bytes.fromhex(_hash(receipt_hash, "receipt_hash"))
    return STRICT_RECEIPT_V2_SIGNATURE_DOMAIN + b"\x00" + _u64(len(key)) + key + digest


def sign_strict_receipt_v2(
    input_value: Any, signer: Any, include_public_key: bool = False
) -> Dict[str, Any]:
    body = build_strict_receipt_v2_body(input_value)
    key_id = strict_receipt_v2_key_id(signer.raw_public_key)
    if body["initiator"]["key_id"] != key_id:
        _fail("signer does not match initiator.key_id")
    receipt_hash = strict_receipt_v2_hash(body)
    preimage = strict_receipt_v2_signature_preimage(key_id, receipt_hash)
    signature = signer.sign_bytes(preimage)
    if (
        not isinstance(signature, str)
        or len(signature) != 128
        or any(character not in _HEX for character in signature)
    ):
        _fail("signer returned an invalid Ed25519 signature")
    public_key_b64 = base64.b64encode(signer.raw_public_key).decode("ascii")
    if signer.public_key_b64 != public_key_b64:
        _fail("signer public_key_b64 does not match raw_public_key")
    from .policy_verify import _resolve_backend

    backend = _resolve_backend()
    try:
        signature_bytes = binascii.unhexlify(signature)
    except (binascii.Error, ValueError) as error:
        raise StrictReceiptV2ValidationError(
            "signer returned an invalid Ed25519 signature"
        ) from error
    if backend is None or not backend(signer.raw_public_key, preimage, signature_bytes):
        _fail("signer signature failed self-verification")
    envelope = {
        "schema": STRICT_RECEIPT_V2_ENVELOPE_SCHEMA,
        "body": body,
        "receipt_hash": receipt_hash,
        "signature": {"algorithm": "Ed25519", "key_id": key_id, "value": signature},
    }
    if include_public_key:
        envelope["public_key_b64"] = public_key_b64
    return envelope

File: sdk-typescript/src/governance/strict-receipt-v2.ts
import { createHash, createPublicKey, verify as cryptoVerify } from 'node:crypto';
import type { DeviceSigner } from '../proxy/device-identity.js';
import { INTENT_V2_ENGINE_VERSION } from '../policy/intent-alignment-v2.js';
import { canonicalJsonForHash } from '../policy/tool-pinning.js';
import type { AarmOutcome } from './aarm-outcome.js';
import {
  STRICT_IDENTIFIER_MAX_BYTES,
  STRICT_SET_MAX_ITEMS,
  boundedCanonicalText,
} from './strict-canonical.js';
import {
  STRICT_RECEIPT_PROFILE_VERSION,
  STRICT_RECEIPT_SCHEMA,
  buildStrictReceiptBody,
  type StrictReceiptBody,
} from './strict-receipt.js';

export const STRICT_RECEIPT_V2_SCHEMA = 'obsvr-strict-receipt-v2' as const;
export const STRICT_RECEIPT_V2_PROFILE_VERSION = '2.0' as const;
export const STRICT_RECEIPT_V2_ENVELOPE_SCHEMA = 'obsvr-strict-receipt-envelope-v2' as const;
export const STRICT_RECEIPT_V2_BODY_DOMAIN = 'obsvr-strict-receipt/body/2' as const;
export const STRICT_RECEIPT_V2_SIGNATURE_DOMAIN = 'obsvr-strict-receipt/signature/2' as const;

const HEX64 = /^[0-9a-f]{64}$/;
const KEY_ID = /^sha256:[0-9a-f]{64}$/;
const SIGNATURE_HEX = /^[0-9a-f]{128}$/;
const ED25519_SPKI_PREFIX = Buffer.from('302a300506032b6570032100', 'hex');

export interface StrictReceiptV2Body {
  schema: typeof STRICT_RECEIPT_V2_SCHEMA;
  profile_version: typeof STRICT_RECEIPT_V2_PROFILE_VERSION;
  record_type: 'decision' | 'resolution';
  receipt_id: string;
  tenant_id: string;
  session_id: string;
  sequence: number;
  timestamp_ms: number;
  clock_regression_clamped: boolean;
  previous_receipt_hash: string | null;
  sdk: StrictReceiptBody['sdk'];
  initiator: StrictReceiptBody['initiator'];
  action: Omit<StrictReceiptBody['action'], 'target'> & { target_hash?: string };
  context: Omit<StrictReceiptBody['context'], 'schema'> & {
    schema: 'obsvr-action-context-v2';
  };
  evaluation: Omit<StrictReceiptBody['evaluation'], 'outcome'> & { outcome: AarmOutcome };
  execution_authorized: boolean;
  suspension?: StrictReceiptBody['suspension'];
  resolution?: StrictReceiptBody['resolution'];
}

export interface StrictReceiptV2Envelope {
  schema: typeof STRICT_RECEIPT_V2_ENVELOPE_SCHEMA;
  body: StrictReceiptV2Body;
  receipt_hash: string;
  signature: { algorithm: 'Ed25519'; key_id: string; value: string };
  public_key_b64?: string;
}

export class StrictReceiptV2ValidationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'StrictReceiptV2ValidationError';
  }
}

function fail(message: string): never { throw new StrictReceiptV2ValidationError(message); }
function record(value: unknown, field: string): Record<string, unknown> {
  if (value === null || typeof value !== 'object' || Array.isArray(value)) {
    return fail(`${field} must be an object`);
  }
  return value as Record<string, unknown>;
}
function exact(value: Record<string, unknown>, allowed: readonly string[], field: string): void {
  const keys = new Set(allowed);
  const unknown = Object.keys(value).filter((key) => !keys.has(key)).sort();
  if (unknown.length) fail(`${field} contains unsupported field: ${unknown[0]}`);
}
function text(value: unknown, field: string): string {
  return boundedCanonicalText(value, field, STRICT_IDENTIFIER_MAX_BYTES, fail);
}
function hash(value: unknown, field: string): string {
  if (typeof value !== 'string' || !HEX64.test(value)) {
    return fail(`${field} must be 64 lowercase hex characters`);
  }
  return value;
}
function optionalHash(
  source: Record<string, unknown>, key: string, field: string,
): string | undefined {
  return Object.prototype.hasOwnProperty.call(source, key) ? hash(source[key], field) : undefined;
}
function boundedTexts(values: readonly string[] | undefined, field: string): void {
  if (values === undefined) return;
  if (values.length > STRICT_SET_MAX_ITEMS) fail(`${field} exceeds ${STRICT_SET_MAX_ITEMS} items`);
  values.forEach((value, index) => text(value, `${field}[${index}]`));
}
function boundedInputTexts(value: unknown, field: string): void {
  if (!Array.isArray(value)) fail(`${field} must be an array`);
  if (value.length > STRICT_SET_MAX_ITEMS) fail(`${field} exceeds ${STRICT_SET_MAX_ITEMS} items`);
  value.forEach((item, index) => text(item, `${field}[${index}]`));
}
function u64(value: number): Buffer {
  const output = Buffer.alloc(8);
  output.writeBigUInt64BE(BigInt(value));
  return output;
}

export function strictReceiptV2KeyId(rawPublicKey: Uint8Array): string {
  const raw = Buffer.from(rawPublicKey);
  if (raw.length !== 32) fail('public key must be 32 raw bytes');
  return `sha256:${createHash('sha256').update(raw).digest('hex')}`;
}

function validateV2TextBounds(body: StrictReceiptV2Body): void {
  [
    [body.receipt_id, 'receipt_id'], [body.session_id, 'session_id'],
    [body.sdk.version, 'sdk.version'], [body.initiator.agent_id, 'initiator.agent_id'],
    [body.action.action_id, 'action.action_id'], [body.action.kind, 'action.kind'],
    [body.action.name, 'action.name'], [body.context.run_id, 'context.run_id'],
    [body.evaluation.engine_version, 'evaluation.engine_version'],
    [body.evaluation.policy_version, 'evaluation.policy_version'],
    [body.evaluation.reason_code, 'evaluation.reason_code'],
  ].forEach(([value, field]) => text(value, field));
  if (body.context.thread_id !== undefined) text(body.context.thread_id, 'context.thread_id');
  boundedTexts(body.evaluation.rule_ids, 'evaluation.rule_ids');
  if (body.suspension) {
    text(body.suspension.suspension_id, 'suspension.suspension_id');
    boundedTexts(body.suspension.required_fields, 'suspension.required_fields');
    if (body.suspension.approval_request_id !== undefined) {
      text(body.suspension.approval_request_id, 'suspension.approval_request_id');
    }
  }
  if (body.resolution) {
    text(body.resolution.suspension_id, 'resolution.suspension_id');
    text(body.resolution.resolver_principal_id, 'resolution.resolver_principal_id');
  }
}

export function buildStrictReceiptV2Body(input: StrictReceiptV2Body): StrictReceiptV2Body {
  const root = record(input, 'receipt body');
  exact(root, [
    'schema', 'profile_version', 'record_type', 'receipt_id', 'tenant_id', 'session_id',
    'sequence', 'timestamp_ms', 'clock_regression_clamped', 'previous_receipt_hash',
    'sdk', 'initiator', 'action', 'context', 'evaluation', 'execution_authorized',
    'suspension', 'resolution',
  ], 'receipt body');
  if (root.schema !== STRICT_RECEIPT_V2_SCHEMA) fail('invalid receipt schema');
  if (root.profile_version !== STRICT_RECEIPT_V2_PROFILE_VERSION) {
    fail('invalid receipt profile_version');
  }
  const tenantId = text(root.tenant_id, 'tenant_id');
  const rawAction = record(root.action, 'action');
  exact(rawAction, ['action_id', 'kind', 'name', 'arguments_hash', 'target_hash',
    'effective_arguments_hash'], 'action');
  const targetHash = optionalHash(rawAction, 'target_hash', 'action.target_hash');
  const rawContext = record(root.context, 'context');
  exact(rawContext, ['schema', 'context_hash', 'run_id', 'thread_id'], 'context');
  if (rawContext.schema !== 'obsvr-action-context-v2') fail('invalid context.schema');
  const rawEvaluation = record(root.evaluation, 'evaluation');
  boundedInputTexts(rawEvaluation.rule_ids, 'evaluation.rule_ids');
  if (Object.prototype.hasOwnProperty.call(root, 'suspension')) {
    const rawSuspension = record(root.suspension, 'suspension');
    boundedInputTexts(rawSuspension.required_fields, 'suspension.required_fields');
  }

  const legacyAction = { ...rawAction };
  delete legacyAction.target_hash;
  if (targetHash !== undefined) legacyAction.target = targetHash;
  const legacyInput: Record<string, unknown> = {
    ...root,
    schema: STRICT_RECEIPT_SCHEMA,
    profile_version: STRICT_RECEIPT_PROFILE_VERSION,
    action: legacyAction,
    context: { ...rawContext, schema: 'obsvr-action-context-v1' },
  };
  delete legacyInput.tenant_id;
  const legacy = buildStrictReceiptBody(legacyInput as unknown as StrictReceiptBody);
  const normalizedAction: StrictReceiptV2Body['action'] = { ...legacy.action };
  delete (normalizedAction as Record<string, unknown>).target;
  if (legacy.action.target !== undefined) normalizedAction.target_hash = legacy.action.target;
  const body: StrictReceiptV2Body = {
    ...legacy,
    schema: STRICT_RECEIPT_V2_SCHEMA,
    profile_version: STRICT_RECEIPT_V2_PROFILE_VERSION,
    tenant_id: tenantId,
    action: normalizedAction,
    context: { ...legacy.context, schema: 'obsvr-action-context-v2' },
  };
  validateV2TextBounds(body);
  if (body.evaluation.engine_version !== INTENT_V2_ENGINE_VERSION) {
    fail(`evaluation.engine_version must be ${INTENT_V2_ENGINE_VERSION}`);
  }
  return body;
}

export function canonicalizeStrictReceiptV2Body(input: StrictReceiptV2Body): string {
  return canonicalJsonForHash(buildStrictReceiptV2Body(input));
}
export function strictReceiptV2Hash(input: StrictReceiptV2Body): string {
  const body = Buffer.from(canonicalizeStrictReceiptV2Body(input), 'utf8');
  return createHash('sha256').update(Buffer.concat([
    Buffer.from(STRICT_RECEIPT_V2_BODY_DOMAIN, 'utf8'), Buffer.from([0]),
    u64(body.length), body,
  ])).digest('hex');
}
export function strictReceiptV2SignaturePreimage(keyId: string, receiptHash: string): Buffer {
  if (!KEY_ID.test(keyId)) fail('invalid strict key id');
  const key = Buffer.from(keyId, 'utf8');
  return Buffer.concat([
    Buffer.from(STRICT_RECEIPT_V2_SIGNATURE_DOMAIN, 'utf8'), Buffer.from([0]),
    u64(key.length), key, Buffer.from(hash(receiptHash, 'receipt_hash'), 'hex'),
  ]);
}

export function signStrictReceiptV2(
  input: StrictReceiptV2Body,
  signer: DeviceSigner,
  includePublicKey = false,
): StrictReceiptV2Envelope {
  const body = buildStrictReceiptV2Body(input);
  const keyId = strictReceiptV2KeyId(signer.rawPublicKey);
  if (body.initiator.key_id !== keyId) fail('signer does not match initiator.key_id');
  const receiptHash = strictReceiptV2Hash(body);
  const preimage = strictReceiptV2SignaturePreimage(keyId, receiptHash);
  const signature = signer.signBytes(preimage);
  if (!SIGNATURE_HEX.test(signature)) fail('signer returned an invalid Ed25519 signature');
  const publicKeyB64 = Buffer.from(signer.rawPublicKey).toString('base64');
  if (signer.publicKeyB64 !== publicKeyB64) fail('signer publicKeyB64 does not match rawPublicKey');
  const publicKey = createPublicKey({
    key: Buffer.concat([ED25519_SPKI_PREFIX, Buffer.from(signer.rawPublicKey)]),
    format: 'der', type: 'spki',
  });
  if (!cryptoVerify(null, preimage, publicKey, Buffer.from(signature, 'hex'))) {
    fail('signer signature failed self-verification');
  }
  const envelope: StrictReceiptV2Envelope = {
    schema: STRICT_RECEIPT_V2_ENVELOPE_SCHEMA,
    body,
    receipt_hash: receiptHash,
    signature: { algorithm: 'Ed25519', key_id: keyId, value: signature },
  };
  if (includePublicKey) envelope.public_key_b64 = publicKeyB64;
  return envelope;
}
